packages/lesswrong/server/repos/SurveySchedulesRepo.ts

```diff
@@ -0,0 +1,109 @@
+import SurveySchedules from "@/lib/collections/surveySchedules/collection";
+import AbstractRepo from "./AbstractRepo";
+import { recordPerfMetrics } from "./perfMetricWrapper";
+
+export type SurveyScheduleWithSurvey = DbSurveySchedule & {
+  survey: DbSurvey,
+};
+
+class SurveySchedulesRepo extends AbstractRepo<"SurveySchedules"> {
+  constructor() {
+    super(SurveySchedules);
+  }
+
+  getCurrentFrontpageSurvey(
+    currentUser: DbUser | null,
+    clientId: string,
+  ): Promise<SurveyScheduleWithSurvey | null> {
+    if (!clientId) {
+      return Promise.resolve(null);
+    }
+    const isLoggedIn = !!currentUser;
+    const userJoin = isLoggedIn
+      ? `LEFT JOIN "Users" u ON u."_id" = $3`
+      : "";
+    const optOutClause = isLoggedIn
+      ? `u."optedOutOfSurveys" IS NOT TRUE AND`
+      : "";
+    const karmaClause = isLoggedIn
+      ? `
+        CASE
+          WHEN ss."minKarma" IS NULL THEN TRUE
+          ELSE ss."minKarma" < COALESCE(u."karma", 0)
+        END AND
+        CASE
+          WHEN ss."maxKarma" IS NULL THEN TRUE
+          ELSE ss."maxKarma" > COALESCE(u."karma", 0)
+        END AND
+      `
+      : ""
+    return this.getRawDb().oneOrNone(`
+      SELECT ss.*, ROW_TO_JSON(s.*) "survey"
+      FROM "SurveySchedules" ss
+      JOIN "Surveys" s ON
+        s."_id" = ss."surveyId"
+      LEFT JOIN "SurveyResponses" sr ON
+        sr."surveyScheduleId" = ss."_id" AND
+        (sr."clientId" = $2 OR (sr."userId" IS NOT NULL AND sr."userId" = $3))
+      ${userJoin}
+      WHERE
+        -- Check the impressions limit hasn't been reached yet
+        (
+          ss."impressionsLimit" IS NULL OR
+          ss."impressionsLimit" > ARRAY_LENGTH(ss."clientIds", 1)
+        ) AND
+        -- Check the user is in the assigned percentage group
+        CASE
+          WHEN ss."maxVisitorPercentage" IS NULL THEN TRUE
+          ELSE ABS(('x' || SUBSTR(MD5(ss."_id" || $2), 1, 16))::BIT(32)::INTEGER) % 100
```

Comment:
I think this should be by `COALESCE(userId, clientId)`, so people aren't re-shown a survey on a different device (this may be covered elsewhere)